plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,144 @@
+                                                          mkPluginHandler)
+import           Language.LSP.Server                     (LspM)
+import           Language.LSP.Types                      (List (List),
+                                                          NormalizedFilePath,
+                                                          Position,
+                                                          ResponseError,
+                                                          SMethod (STextDocumentSelectionRange),
+                                                          SelectionRange (..),
+                                                          SelectionRangeParams (..),
+                                                          TextDocumentIdentifier (TextDocumentIdentifier),
+                                                          Uri)
+import           Prelude                                 hiding (span)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    liftIO $ logInfo logger $ "requesting selection range for file: " <> T.pack (show uri)
+    response $ do
+        filePath <- ExceptT . pure . maybeToEither "fail to convert uri to file path" $
+                toNormalizedFilePath' <$> uriToFilePath' uri
+        selectionRanges <- ExceptT . liftIO . runIdeAction "SelectionRange" (shakeExtras ide) . runExceptT $
+            getSelectionRanges filePath positions
+        pure . List $ selectionRanges
+  where
+    uri :: Uri
+    TextDocumentIdentifier uri = _textDocument
+
+    positions :: [Position]
+    List positions = _positions
+
+    logger = ideLogger ide
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> ExceptT String IdeAction [SelectionRange]
+getSelectionRanges file positions = do
+    (HAR{hieAst, refMap}, positionMapping) <- maybeToExceptT "fail to get hie ast" $ useE GetHieAst file
+    -- 'positionMapping' should be applied to the input positions before using them
+    positions' <- maybeToExceptT "fail to apply position mapping to input positions" . MaybeT . pure $
+        traverse (fromCurrentPosition positionMapping) positions
+
+    ast <- maybeToExceptT "fail to get ast for current file" . MaybeT . pure $
+        -- in GHC 9, the 'FastString' in 'HieASTs' is replaced by a newtype wrapper around 'LexicalFastString'
+        -- so we use 'coerce' to make it work in both GHC 8 and 9
+        getAsts hieAst Map.!? (coerce . mkFastString . fromNormalizedFilePath) file
+
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    let selectionRanges = findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
```

Comment:
`(astPathsLeafToRoot ast'` is expensive, so we should cache its result and avoid doing it on every call. Please consider moving it to a new build rule `GetSelectionRanges`.

Reply:
For performance issues, I will benchmark for large files in a future PR. As at least this action works without any delay in a file with 1000 lines on my local machine (intel 11th gen chip, not a fast one), so maybe it's ok for now?

Reply:
Do we understand at least how often the LSP client requests the selection ranges? Is it done on cursor move, or only when the user requests a selection expand?

Reply:
> Do we understand at least how often the LSP client requests the selection ranges? Is it done on cursor move, or only when the user requests a selection expand?

Only when the user requests a selection expand. Even lesser frequent than that, in fact. If a user start expanding/shrinking from one position, and doesn't change the selection area manually, then this lsp method will only be called once. (I think that's why each `SelectionRange` is a linked list)

Reply:
Ok, if this is the case then the need to optimize is much less pressing.